NullPointerException: Cannot invoke "org.h2.mvstore.tx.Transaction.getStatus()" because "this.transaction" is null

This happened in 2.1.214 and also 2.4.240, under some (not too high) concurrency, we often get this NPE:
```
org.jooq.exception.DataAccessException: Cannot get transactionIsolation
	at org.jooq.impl.DefaultConnectionProvider.getTransactionIsolation(DefaultConnectionProvider.java:270)
	at org.jooq.impl.DefaultTransactionProvider.isolation(DefaultTransactionProvider.java:156)
	at org.jooq.impl.DefaultTransactionProvider.brace(DefaultTransactionProvider.java:286)
	at org.jooq.impl.DefaultTransactionProvider.begin(DefaultTransactionProvider.java:203)
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$3(DefaultDSLContext.java:559)
	at org.jooq.impl.Tools$3$1.block(Tools.java:6543)
	at java.base/java.util.concurrent.ForkJoinPool.unmanagedBlock(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.managedBlock(Unknown Source)
	at org.jooq.impl.Tools$3.get(Tools.java:6540)
	at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:613)
	at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:522)
	...
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.tx.Transaction.getStatus()"" because ""this.transaction"" is null"; SQL statement:
SELECT ISOLATION_LEVEL FROM INFORMATION_SCHEMA.SESSIONS WHERE SESSION_ID = SESSION_ID() [50000-240]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:566)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
	at org.h2.message.DbException.get(DbException.java:212)
	at org.h2.message.DbException.convert(DbException.java:407)
	at org.h2.command.Command.executeQuery(Command.java:219)
	at org.h2.server.TcpServerThread.process(TcpServerThread.java:360)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:193)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException: Cannot invoke "org.h2.mvstore.tx.Transaction.getStatus()" because "this.transaction" is null
	at org.h2.engine.SessionLocal.hasPendingTransaction(SessionLocal.java:867)
	at org.h2.table.InformationSchemaTable.sessions(InformationSchemaTable.java:2711)
	at org.h2.table.InformationSchemaTable.sessions(InformationSchemaTable.java:2675)
	at org.h2.table.InformationSchemaTable.generateRows(InformationSchemaTable.java:980)
	at org.h2.index.MetaIndex.find(MetaIndex.java:53)
```

It happens since jooq always reads the transaction isolation on every request, leading to the query "SELECT ISOLATION_LEVEL FROM INFORMATION_SCHEMA.SESSIONS". There seems to be a race condition if there are concurrent writing requests while this meta table is built:
Building that meta-table makes H2 iterate over <b>all</b> live sessions and call `SessionLocal.hasPendingTransaction()` on each (a non-atomic `transaction != null` check followed by `transaction.getStatus()`. When another session commits (nulling its `transaction`) in between, the read races and H2 throws `NullPointerException`.

It can easily be reproduced by having one thread repeatedly accessing the transaction isolation level while another repeatedly updates some data.

See this unit test demonstrating the bug:
[H2SessionsIsolationRaceTest.java](https://github.com/user-attachments/files/29100534/H2SessionsIsolationRaceTest.java)

--- a/h2/src/main/org/h2/engine/SessionLocal.java
+++ b/h2/src/main/org/h2/engine/SessionLocal.java
@@ -551,8 +551,9 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
 
     public void setLockTimeout(int lockTimeout) {
         this.lockTimeout = lockTimeout;
-        if (hasTransaction()) {
-            transaction.setTimeoutMillis(lockTimeout);
+        Transaction tx = transaction;
+        if (tx != null) {
+            tx.setTimeoutMillis(lockTimeout);
         }
     }
 
@@ -685,10 +686,11 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
      */
     public void commit(boolean ddl) {
         beforeCommitOrRollback();
-        if (hasTransaction()) {
+        Transaction tx = transaction;
+        if (tx != null) {
             try {
                 markUsedTablesAsUpdated();
-                transaction.commit();
+                tx.commit();
                 markUsedTablesAsUpdated();
                 removeTemporaryLobs(true);
                 endTransaction();
@@ -794,7 +796,8 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
             getDatabase().releaseDatabaseObjectIds(idsToRelease);
             idsToRelease = null;
         }
-        if (hasTransaction() && !transaction.allowNonRepeatableRead()) {
+        Transaction tx = transaction;
+        if (tx != null && !tx.allowNonRepeatableRead()) {
             snapshotDataModificationId = getDatabase().getNextModificationDataId();
         }
     }
@@ -814,9 +817,7 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
      */
     public void rollback() {
         beforeCommitOrRollback();
-        if (hasTransaction()) {
-            rollbackTo(null);
-        }
+        rollbackTo(null);
         idsToRelease = null;
         cleanTempTables(false);
         if (autoCommitAtTransactionEnd) {
@@ -833,13 +834,14 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
      */
     public void rollbackTo(Savepoint savepoint) {
         int index = savepoint == null ? 0 : savepoint.logIndex;
-        if (hasTransaction()) {
+        Transaction tx = transaction;
+        if (tx != null) {
             markUsedTablesAsUpdated();
             if (savepoint == null) {
-                transaction.rollback();
+                tx.rollback();
                 transaction = null;
             } else {
-                transaction.rollbackToSavepoint(savepoint.transactionSavepoint);
+                tx.rollbackToSavepoint(savepoint.transactionSavepoint);
             }
             markUsedTablesAsUpdated();
         }
@@ -855,7 +857,7 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
         }
 
         // Because cache may have captured query result (in Query.lastResult),
-        // which is based on data from uncommitted transaction.,
+        // which is based on data from uncommitted transaction.
         // It is not valid after rollback, therefore cache has to be cleared.
         if (queryCache != null) {
             queryCache.clear();
@@ -864,7 +866,8 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
 
     @Override
     public boolean hasPendingTransaction() {
-        return containsUncommitted() && transaction.getStatus() != Transaction.STATUS_PREPARED;
+        Transaction tx = transaction;
+        return containsUncommitted(tx) && tx.getStatus() != Transaction.STATUS_PREPARED;
     }
 
     /**
@@ -979,10 +982,6 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
         locks.remove(t);
     }
 
-
-    private boolean hasTransaction() {
-        return transaction != null;
-    }
 
     private void unlockAll() {
         if (!locks.isEmpty()) {
@@ -1115,12 +1114,13 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
     }
 
     /**
-     * Whether the session contains any uncommitted changes.
+     * Whether the transaction contains any uncommitted changes.
+     * @param tx to check
      *
      * @return true if yes
      */
-    public boolean containsUncommitted() {
-        return hasTransaction() && transaction.hasChanges();
+    private static boolean containsUncommitted(Transaction tx) {
+        return tx != null && tx.hasChanges();
     }
 
     /**
@@ -1608,10 +1608,11 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
     }
 
     public Value getTransactionId() {
-        if (!containsUncommitted()) {
+        Transaction tx = transaction;
+        if (!containsUncommitted(tx)) {
             return ValueNull.INSTANCE;
         }
-        return ValueVarchar.get(Long.toString(transaction.getSequenceNum()));
+        return ValueVarchar.get(Long.toString(tx.getSequenceNum()));
     }
 
     /**
@@ -1729,8 +1730,9 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
      */
     public void endStatement() {
         setCurrentCommand(null);
-        if (hasTransaction()) {
-            transaction.markStatementEnd();
+        Transaction tx = transaction;
+        if (tx != null) {
+            tx.markStatementEnd();
         }
         startStatement = -1;
         statementModificationDataId = 0L;
@@ -1796,7 +1798,8 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
     }
 
     public int getBlockingSessionId() {
-        return hasTransaction() ? transaction.getBlockerId() : 0;
+        Transaction tx = transaction;
+        return tx != null ? tx.getBlockerId() : 0;
     }
 
     @Override

--- a/h2/src/main/org/h2/table/InformationSchemaTableLegacy.java
+++ b/h2/src/main/org/h2/table/InformationSchemaTableLegacy.java
@@ -1775,12 +1775,17 @@ public final class InformationSchemaTableLegacy extends MetaTable {
                     NetworkConnectionInfo networkConnectionInfo = s.getNetworkConnectionInfo();
                     Command command = s.getCurrentCommand();
                     int blockingSessionId = s.getBlockingSessionId();
+                    User user = s.getUser();
+                    if (user == null) {
+                        // Session was closed concurrently
+                        continue;
+                    }
                     add(session,
                             rows,
                             // ID
                             ValueInteger.get(s.getId()),
                             // USER_NAME
-                            s.getUser().getName(),
+                            user.getName(),
                             // SERVER
                             networkConnectionInfo == null ? null : networkConnectionInfo.getServer(),
                             // CLIENT_ADDR

--- a/h2/src/test/org/h2/test/jdbc/TestConcurrentConnectionUsage.java
+++ b/h2/src/test/org/h2/test/jdbc/TestConcurrentConnectionUsage.java
@@ -9,6 +9,7 @@ import java.io.ByteArrayInputStream;
 import java.sql.Connection;
 import java.sql.PreparedStatement;
 import java.sql.SQLException;
+import java.sql.Statement;
 
 import org.h2.test.TestBase;
 import org.h2.test.TestDb;
@@ -31,6 +32,8 @@ public class TestConcurrentConnectionUsage extends TestDb {
     @Override
     public void test() throws SQLException {
         testAutoCommit();
+        testSessionUsage(false);
+        testSessionUsage(true);
     }
 
     private void testAutoCommit() throws SQLException {
@@ -56,4 +59,36 @@ public class TestConcurrentConnectionUsage extends TestDb {
         conn.close();
     }
 
+    
+    private void testSessionUsage(boolean useLegacy) throws SQLException {
+        deleteDb(getTestName());
+        try (Connection infoConn = getConnection(getTestName() + ";OLD_INFORMATION_SCHEMA=" + useLegacy)) {
+            Statement stmt = infoConn.createStatement();
+            stmt.execute("CREATE TABLE t(id INT PRIMARY KEY)");
+
+            int threadCount = 8;
+            Task[] tasks = new Task[threadCount];
+            for (int i = 0; i < threadCount; i++) {
+                tasks[i] = new Task() {
+                    @Override
+                    public void call() throws Exception {
+                        try (Connection txConn = getConnection(getTestName())) {
+                            Statement st = txConn.createStatement();
+                            while (!stop) {
+                                st.executeUpdate("MERGE INTO t(id) KEY (id) VALUES (0)");
+                            }
+                        }
+                    }
+                }.execute();
+            }
+
+            for (int i = 0; i < 1_000; i++) {
+                stmt.executeQuery("SELECT ISOLATION_LEVEL FROM INFORMATION_SCHEMA.SESSIONS").close();
+            }
+
+            for (Task task : tasks) {
+                task.get();
+            }
+        }
+    }
 }

